app/app.go

```diff
@@ -2,26 +2,102 @@ package app
 
 import (
 	"fmt"
+	"io"
+	"os"
+	"path"
+
 	. "github.com/tendermint/go-common"
+	dbm "github.com/tendermint/go-db"
 	"github.com/tendermint/go-merkle"
 	"github.com/tendermint/go-wire"
 	tmsp "github.com/tendermint/tmsp/types"
 )
 
 type MerkleEyesApp struct {
-	tree merkle.Tree
+	tree       merkle.Tree
+	db         dbm.DB
+	persistent bool
+}
+
+//Database Keys
+var saveKey []byte = []byte{0x00} //Key for merkle tree save value db values
+var prefix []byte = []byte{0x01}  //Prefix byte for all data saved in the merkle tree, used to prevent key collision with the savekey record
+
+//App Usage Keys
+const WriteSet byte = 0x01
+const WriteRem byte = 0x02
+const ReadByKey byte = 0x01
+const ReadByIndex byte = 0x02
+const ReadSize byte = 0x03
+
+func NewMerkleEyesApp(dbName string, cache int) *MerkleEyesApp {
+
+	//non-persistent case
+	if dbName == "" {
+		tree := merkle.NewIAVLTree(
+			0,
+			nil,
+		)
+		return &MerkleEyesApp{
+			tree:       tree,
+			db:         nil,
+			persistent: false,
+		}
+	}
+
+	//setup the persistent merkle tree
+	present, _ := IsDirEmpty(path.Join(dbName, dbName+".db"))
+
+	//open the db, if the db doesn't exist it will be created
+	db := dbm.NewDB(dbName, dbm.DBBackendLevelDB, dbName)
+
+	// Load Tree
+	tree := merkle.NewIAVLTree(cache, db)
+
+	if present {
+		fmt.Println("no existing db, creating new db")
+		db.Set(saveKey, tree.Save())
+	} else {
+		fmt.Println("loading existing db")
+	}
+
+	//load merkle state
+	tree.Load(db.Get(saveKey))
+
+	return &MerkleEyesApp{
+		tree:       tree,
+		db:         db,
+		persistent: true,
+	}
+}
+
+//append the prefix before the key, used to prevent db collisions
+func AddPrefix(key []byte) []byte {
+	return append(prefix, key...)
+}
+
+func IsDirEmpty(name string) (bool, error) {
```

Comment:
Agreed, maybe this can be added to go-common? Otherwise I could define a common package within merkleeyes. Thoughts?